src/Data/BTree/Insert.hs

```diff
@@ -104,4 +139,81 @@ insertTree key val tree
               , treeRootId = Just newRootId
               }
 
+insertTreeMany :: (AllocM m, Key key, Value val)
+    => Map key val
+    -> Tree key val
+    -> m (Tree key val)
+insertTreeMany kvs tree
+    | Tree
+      { treeHeight = height
+      , treeRootId = Just rootId
+      } <- tree
+    = do
+        newRootIdx <- insertRecMany kvs height rootId
+        case fromSingletonIndex newRootIdx of
+            Just newRootId ->
+                fixUp $! Tree
+                    { treeHeight = height
+                    , treeRootId = Just newRootId
+                    }
+            Nothing -> do
+                let newHeight = incrHeight height
+                newRootId <- allocNode newHeight Idx
+                    { idxChildren = newRootIdx }
+                fixUp $! Tree
+                    { treeHeight = newHeight
+                    , treeRootId = Just newRootId
+                    }
+    | Tree { treeRootId = Nothing } <- tree
+    = do
+        newRootId <- allocNode zeroHeight Leaf { leafItems = kvs }
+        fixUp $! Tree
+            { treeHeight = zeroHeight
+            , treeRootId = Just newRootId
+            }
+
+{-| Fix up the root node of a tree.
+
+    Fix up the root node of a tree, where all other nodes are valid, but the
+    root node may contain more items than allowed. Do this by repeatedly
+    splitting up the root node.
+-}
+fixUp :: (AllocM m, Key k, Value v) => Tree k v -> m (Tree k v)
```

Comment:
So here is actually where you can't really misuse the Tree datatype. You shouldn't really store too large Nodes. Once we go to binary serialization this probably needs to be reimplemented.